include/qore/intern/qore_thread_intern.h

```diff
@@ -418,40 +412,98 @@ class QoreParseClassHelper {
    DLLLOCAL ~QoreParseClassHelper();
 };
 
-class QoreProgramLocationHelper {
+class QoreProgramStackLocationHelper {
+public:
+    DLLLOCAL QoreProgramStackLocationHelper(QoreStackLocation* stack_loc, const AbstractStatement*& current_stmt,
+        QoreProgram*& current_pgm) :
+        stack_loc(update_get_runtime_stack_location(stack_loc, current_stmt, current_pgm)) {
+    }
+
+    DLLLOCAL ~QoreProgramStackLocationHelper() {
+        update_runtime_stack_location(stack_loc);
+    }
+
 protected:
-   const QoreProgramLocation* loc;
-   const AbstractStatement* statement;
+    const QoreStackLocation* stack_loc;
+};
+
+class QoreInternalCallStackLocationHelperBase : public QoreStackLocation, public QoreProgramStackLocationHelper {
 public:
-   DLLLOCAL QoreProgramLocationHelper(const QoreProgramLocation* n_loc, const AbstractStatement* n_stat = nullptr) : loc(update_get_runtime_location(n_loc)), statement(update_get_runtime_statement(n_stat)) {
-   }
+    DLLLOCAL QoreInternalCallStackLocationHelperBase() : QoreProgramStackLocationHelper(this, stmt, pgm) {
+    }
 
-   DLLLOCAL ~QoreProgramLocationHelper() {
-      update_runtime_location(loc);
-      update_get_runtime_statement(statement);
-   }
+    DLLLOCAL virtual QoreProgram* getProgram() const {
+        return pgm;
+    }
+
+    DLLLOCAL virtual const AbstractStatement* getStatement() const {
+        return stmt;
+    }
+
+protected:
+    const AbstractStatement* stmt;
+    QoreProgram* pgm;
 };
 
-class QoreProgramOptionalLocationHelper {
+class QoreInternalCallStackLocationHelper : public QoreInternalCallStackLocationHelperBase {
+public:
+    DLLLOCAL QoreInternalCallStackLocationHelper(const QoreProgramLocation& loc, const char* call,
+        qore_call_t call_type) : loc(loc), call(call), call_type(call_type) {
```

Comment:
`call` `const char *` is stored into a member but who is the owner and how it is forced that the pointer is going to be valid as long as this instance lives?

Reply:
done